core/src/main/java/com/yahoo/oak/HashChunk.java

```diff
@@ -0,0 +1,413 @@
+
+        protected void setIsEndBoundCheckNeeded(
+            ThreadContext ctx, K to, boolean toInclusive, OakScopedReadBuffer chunkBoundaryKey) {
+            this.endBound = to;
+            this.endBoundInclusive = toInclusive;
+
+            if (this.endBound == null || !isKeyOutOfEndBound(chunkBoundaryKey)) {
+                // isEndBoundCheckNeeded is NEVER_END_BOUNDARY_CHECK by default
+                return;
+            }
+
+            // generally there is a need for the boundary check, but maybe delay it to the middle
+            // of the chunk. isEndBoundCheckNeeded value can still be changed
+            isEndBoundCheckNeeded = IterEndBoundCheck.ALWAYS_END_BOUNDARY_CHECK;
+        }
+    }
+
+    /********************************************************************************************/
+    /*------------------------------- Iterators Implementations --------------------------------*/
+    class AscendingIter extends ChunkIter {
+
+        AscendingIter(ThreadContext ctx, K to, boolean toInclusive,
+            OakScopedReadBuffer nextChunkMinKey) {
+            next = 0;
+            next = advanceNextIndexNoBound(next, ctx);
+            setIsEndBoundCheckNeeded(ctx, to, toInclusive, nextChunkMinKey);
+        }
+
+        AscendingIter(ThreadContext ctx, K from, boolean fromInclusive, K to, boolean toInclusive,
+            OakScopedReadBuffer nextChunkMinKey) {
+            KeyBuffer tempKeyBuff = ctx.tempKey;
+            next = 0;
+        }
+
+        private void advance(ThreadContext ctx) {
+            next++;
+            // if there is no need to check the end-boundary on this chunk (IterEndBoundCheck.NEVER_END_BOUNDARY_CHECK),
+            // or if the caller will check the end-boundary (IterEndBoundCheck.ALWAYS_END_BOUNDARY_CHECK),
+            // then advance next without additional checks
+            if (isEndBoundCheckNeeded != IterEndBoundCheck.MID_END_BOUNDARY_CHECK) {
+                advanceNextIndexNoBound(next, ctx);
+            } else {
+                next = advanceNextIndex(next, ctx);
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return true;
+        }
+
+        @Override
+        public int next(ThreadContext ctx) {
+            int toReturn = next;
+            advance(ctx);
+            return toReturn;
+        }
+
+        private int advanceNextIndex(final int entryIndex, ThreadContext ctx) {
+            int next = entryIndex;
+            while (!entryHashSet.isValueRefValidAndNotDeleted(next)) {
+                next++;
+            }
+            return next;
+        }
+
+        private int advanceNextIndexNoBound(final int entryIndex, ThreadContext ctx) {
+            int next = entryIndex;
+            while (!entryHashSet.isValueRefValidAndNotDeleted(next)) {
+                next++;
+            }
+            return next;
+        }
+
+        @Override
+        protected boolean isKeyOutOfEndBound(OakScopedReadBuffer key) {
```

Comment:
In general, I didn't take care for iterators code at all. Just left it there, so maybe it can be taken as a basic code when we will write iterators for HashChunk. So I am leaving everything related for later.